File: src/main/java/math/Matrix4d.java
package math;

import org.jetbrains.annotations.NotNull;

public class Matrix4d extends MatrixBuilder{

    private float[][] matrix;

    public Matrix4d() {

    }

    public Matrix4d(float[][] m) {
        this.matrix = m;
    }

    public float[][] getMatrix4() {
        return matrix;
    }

    public void setMatrix4(float[][] m) {
        this.matrix = m;
    }

    public static Matrix4d identity4() {
        Matrix4d im = new Matrix4d();
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < 4; j++)
                if (i == j)
                    im.matrix[i][j] = 1;
                else
                    im.matrix[i][j] = 0;
        return im;
    }

    public void setValue(float value, int x, int y){
        this.matrix[x][y] = value;
    }

    public void setRow(int rowIndex, float[] row){
        for(int i = 0; i < 4; i++)
            for(int j = 0; j < 4; j++)
                if(i == rowIndex)
                    this.matrix[i][j] = row[j];
    }

    public void setColumn(int columnIndex, float[] column){
        for(int i = 0; i < 4; i++)
            for(int j = 0; j < 4; j++)
                if(j == columnIndex)
                    this.matrix[i][j] = column[i];
    }

    @Override
    public void plusAssign(double scalar) {

    }

    @Override
    public void minusAssign(double scalar) {

    }

    @NotNull
    @Override
    public MatrixBuilder plus(double scalar) {
        return null;
    }

    @NotNull
    @Override
    public MatrixBuilder minus(double scalar) {
        return null;
    }

    @NotNull
    @Override
    public MatrixBuilder inc() {
        return null;
    }

    @NotNull
    @Override
    public MatrixBuilder dec() {
        return null;
    }
}

File: src/main/java/math/Matrix4d.kt
package math

class Matrix4d(
    private var array: Array<Array<Double>> = Array(4) { Array(4) { .0 } }
) : MatrixBuilder(array) {

    // auxiliary constructor
    constructor(value: Double) : this(
        Array(4) { Array(4) { value } }
    )

    // getters and setters
    fun getArray() = array

    infix fun setArray(arr: Array<Array<Double>>) {
        array = arr
    }

    fun setValue(value: Double, x: Int, y: Int) {
        if (areCoordinatesInBoundaries(x, y))
            array[x][y] = value
    }

    fun getValue(x: Int, y: Int) {
        if (areCoordinatesInBoundaries(x, y))
            array[x][y]
    }

    fun setRow(rowIndex: Int, row: DoubleArray) {
        for (i in 0..3) for (j in 0..3) if (i == rowIndex) array[i][j] = row[j]
    }

    fun setColumn(columnIndex: Int, column: DoubleArray) {
        for (i in 0..3) for (j in 0..3) if (j == columnIndex) array[i][j] = column[i]
    }

    // abstract members' implementation
    override fun plusAssign(scalar: Double) {
        this += Matrix4d(scalar)
    }

    override fun minusAssign(scalar: Double) {
        this -= Matrix4d(scalar)
    }

    override fun plus(scalar: Double) = (this + Matrix4d(scalar))!!

    override fun minus(scalar: Double) = (this - Matrix4d(scalar))!!

    override fun inc() = this + 1.0

    override fun dec() = this - 1.0

    // static instances
    companion object {
        private val _IDENTITY_ARRAY by lazy {
            val result = Array(4) { Array(4) { .0 } }
            for (i in 0 until 4)
                result[i][i] = 1.0
            result
        }

        @JvmStatic
        val IDENTITY_MATRIX: Matrix4d
            get() = Matrix4d(_IDENTITY_ARRAY)
    }
}